system-addon/lib/LinksCache.jsm

```diff
@@ -0,0 +1,101 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+this.EXPORTED_SYMBOLS = ["LinksCache"];
+
+const EXPIRATION_TIME = 5 * 60 * 1000; // 5 minutes
+
+/**
+ * Cache link results from a provided object property and refresh after some
+ * amount of time has passed. Allows for migrating data from previously cached
+ * links to the new links with the same url.
+ */
+this.LinksCache = class LinksCache {
+
+  /**
+   * Create a links cache for a given object property.
+   *
+   * @param {object} linkObject Object containing the link property
+   * @param {string} linkProperty Name on object to access
+   * @param {function} migrator Optional callback receiving the old and new link
+   * @param {function} shouldRefresh Optional callback receiving the old and new options
+   */
+  constructor(linkObject, linkProperty, migrator = () => {}, shouldRefresh = () => {}) {
+    this.clear();
+    this.expire();
+    // Allow getting links from both methods and array properties
+    this.linkGetter = options => {
+      const ret = linkObject[linkProperty];
+      return typeof ret === "function" ? ret.call(linkObject, options) : ret;
+    };
+    this.migrator = migrator;
+    this.shouldRefresh = shouldRefresh;
+  }
+
+  /**
+   * Clear the cached data.
+   */
+  clear() {
+    this.cache = Promise.resolve([]);
+    this.lastOptions = {};
+  }
+
+  /**
+   * Force the next request to update the cache.
+   */
+  expire() {
+    // Pretend the we expired at some negative time (tests start at T=0)
+    this.lastUpdate = 0 - EXPIRATION_TIME - 1;
```

Comment:
Personally I would change `request` to check `if (!this.lastUpdate || now > this.lastUpdate + EXPIRATION_TIME`) and have this method just delete `this.lastUpdate` to avoid the weird math there

Reply:
This would also prevent having to explicitly call `expire()` in the constructor, since `lastUpdate` is only set by `request`.

Reply:
Okay switching to `delete` although I just realized we actually want to call `expire` from `clear` to correctly forget the last update time.